File: bench/stress/s4_match.go
// S4 — pattern matching, Go (switch-true mirrors the arm order). go run s4_match.go
package main

import "fmt"

func weight(m int64) int64 {
	switch {
	case m == 0:
		return 7
	case m >= 1 && m <= 3:
		return 2
	case m == 11:
		return 5
	case m > 7:
		return 3
	default:
		return 1
	}
}

func main() {
	var n int64 = 20_000_000
	var s int64
	for k := int64(0); k < n; k++ {
		s += weight(k % 12)
	}
	fmt.Println(s)
}

File: bench/stress/s4_match.rs
// S4 — pattern matching, Rust. rustc -O s4_match.rs -o s4_match
fn weight(m: i64) -> i64 {
    match m {
        0 => 7,
        1 | 2 | 3 => 2,
        11 => 5,
        x if x > 7 => 3,
        _ => 1,
    }
}

fn main() {
    let n: i64 = 20_000_000;
    let mut s: i64 = 0;
    for k in 0..n {
        s += weight(k % 12);
    }
    println!("{}", s);
}
